OpenRA.Mods.Common/Traits/Render/DrawLineToTarget.cs

```diff
@@ -62,80 +44,62 @@ void INotifySelected.Selected(Actor a)
 			lifetime = info.Delay;
 		}
 
+		void INotifySelected.Selected(Actor a)
+		{
+			ShowTargetLines(a);
+		}
+
 		IEnumerable<IRenderable> IRenderAboveShroudWhenSelected.RenderAboveShroud(Actor self, WorldRenderer wr)
 		{
-			var force = Game.GetModifierKeys().HasModifier(Modifiers.Alt);
-			if ((lifetime <= 0 || --lifetime <= 0) && !force)
+			if (self.Owner != self.World.LocalPlayer)
 				yield break;
 
-			if (!(force || Game.Settings.Game.DrawTargetLine))
+			// shift is "force" too, players want to see the lines when in waypoint mode.
+			bool force = Game.GetModifierKeys().HasModifier(Modifiers.Alt)
+			          || Game.GetModifierKeys().HasModifier(Modifiers.Shift);
+
+			if ((lifetime <= 0 || --lifetime <= 0) && !force)
 				yield break;
 
-			if (targets == null || targets.Count == 0)
+			if (!(force || Game.Settings.Game.DrawTargetLine))
 				yield break;
 
-			foreach (var target in targets)
+			WPos prev = self.CenterPosition;
+			for (var a = self.CurrentActivity; a != null; a = a.NextActivity)
 			{
-				if (target.Type == TargetType.Invalid)
-					continue;
-
-				yield return new TargetLineRenderable(new[] { self.CenterPosition, target.CenterPosition }, c);
+				if (a is OpenRA.Mods.Common.Activities.Move.MovePart)
+					a = ((OpenRA.Mods.Common.Activities.Move.MovePart)a).Move;
+
+				foreach (var target in a.GetTargets(self))
+				{
+					if (a.IsCanceled || target.Type == TargetType.Invalid)
+						continue;
+
+					yield return new TargetLineRenderable(new[] { prev, target.CenterPosition }, a.TargetLineColor);
+					prev = target.CenterPosition;
+				}
+
+				if (a is ResupplyAircraft)
```

Comment:
I see, you want to use TargetLineNode as a return value so that everything needed is returned in one shot.